File: frontend/src/gameArena/ecs/entities/ClosingZoneSystem.js
import {
    CLOSING_ZONE_CONFIG,
    CLOSING_ZONE_ID,
    CLOSING_ZONE_TYPE,
} from "../../gameconfig/ArenaHazardConfig.js";

/**
 * The closing zone is a world hazard rather than a bot-visible ability
 * entity. Its replay shape is still entity-like so the renderer and replay
 * pipeline can reuse the existing arena shape architecture.
 */
export function isClosingZone(shape) {
    return shape?.type === CLOSING_ZONE_TYPE || shape?.id === CLOSING_ZONE_ID;
}

export function createClosingZone(width, height) {
    const safeRadius = fullArenaCircleRadius(width, height);
    return zoneShape(width, height, safeRadius, 0, 0);
}

export function closingZoneSafeRadius(elapsedMs, width, height, config = CLOSING_ZONE_CONFIG) {
    const activeElapsedMs = Math.max(0, Number(elapsedMs) - Number(config.startDelayMs));
    if (Number(elapsedMs) < Number(config.startDelayMs)) return null;

    const fullRadius = fullArenaCircleRadius(width, height, config);
    const phaseCount = Math.max(1, Math.floor(Number(config.phaseCount) || 1));
    const approachMs = Math.max(0, Number(config.approachDurationMs) || 0);
    const phaseMs = Math.max(1, Number(config.phaseDurationMs) || 1);
    const completionElapsedMs = approachMs + Math.max(0, phaseCount - 1) * phaseMs;
    const phaseElapsedMs = activeElapsedMs % phaseMs;
    if (activeElapsedMs >= completionElapsedMs) return 0;
    if (phaseElapsedMs <= approachMs) {
        const phaseIndex = Math.floor(activeElapsedMs / phaseMs);
        const progress = approachMs > 0 ? phaseElapsedMs / approachMs : 1;
        const startFraction = Math.max(0, 1 - phaseIndex / phaseCount);
        const endFraction = Math.max(0, startFraction - 1 / phaseCount);
        return fullRadius * lerp(startFraction, endFraction, progress);
    }

    // The remaining portion of each configured phase is a hold: the safe
    // radius and hitbox stay cached while damage continues for bots outside it.
    const phaseIndex = Math.floor(activeElapsedMs / phaseMs);
    const startFraction = Math.max(0, 1 - phaseIndex / phaseCount);
    if (phaseIndex >= phaseCount) return 0;
    return fullRadius * Math.max(0, startFraction - 1 / phaseCount);
}

export function closingZoneTouchesBot(zone, bot) {
    if (!zone || !bot || Number(bot.hp ?? 0) <= 0) return false;
    const safeRadius = Number(zone.safeRadius ?? Number(zone.size ?? 0) / 2);
    const botRadius = Math.max(0, Number(bot.size ?? 0)) / 2;
    const distance = Math.hypot(
        Number(bot.x ?? 0) - Number(zone.x ?? 0),
        Number(bot.y ?? 0) - Number(zone.y ?? 0),
    );
    return distance + botRadius >= safeRadius;
}

export function tickClosingZoneWorld({
    zone,
    bots,
    elapsedMs,
    stepMs = 100,
    width,
    height,
} = {}, { applyDamageToShape } = {}) {
    const currentElapsedMs = Math.max(0, Number(elapsedMs) || 0);
    if (currentElapsedMs < Number(CLOSING_ZONE_CONFIG.startDelayMs)) {
        return { bots, zone: null };
    }

    const activeElapsedMs = Math.max(0, currentElapsedMs - CLOSING_ZONE_CONFIG.startDelayMs);
    const currentZone = zone ?? createClosingZone(width, height);
    const updateIntervalMs = Math.max(1, Number(CLOSING_ZONE_CONFIG.geometryUpdateMs) || 1);
    const geometryElapsedMs = Math.floor(activeElapsedMs / updateIntervalMs) * updateIntervalMs;
    const candidateSafeRadius = closingZoneSafeRadius(
        Number(CLOSING_ZONE_CONFIG.startDelayMs) + geometryElapsedMs,
        width,
        height,
    );
    const geometryChanged = Math.abs(candidateSafeRadius - Number(currentZone.safeRadius ?? 0)) > 1e-9;
    const nextZone = geometryChanged
        ? updateZoneGeometry(currentZone, width, height, geometryElapsedMs, candidateSafeRadius)
        : currentZone;

    let nextBots = bots ?? [];
    const previousElapsedMs = Math.max(0, currentElapsedMs - Math.max(1, Number(stepMs) || 1));
    const damageTickCount = completedDamageTicks(previousElapsedMs, currentElapsedMs);
    if (damageTickCount > 0 && typeof applyDamageToShape === "function") {
        const damageSource = { id: CLOSING_ZONE_ID, type: CLOSING_ZONE_TYPE };
        for (let tick = 0; tick < damageTickCount; tick += 1) {
            nextBots = nextBots.map((bot) => {
                if (!closingZoneTouchesBot(nextZone, bot)) return bot;
                const damage = Math.max(0, roundCombatValue(
                    Number(bot.maxHp ?? bot.hp ?? 0) * Number(CLOSING_ZONE_CONFIG.damagePercentMaxHp),
                ));
                return applyDamageToShape(bot, damage, damageSource);
            });
        }
    }

    return { bots: nextBots, zone: nextZone };
}

function roundCombatValue(value) {
    return Math.round(Number(value) * 1000) / 1000;
}

function updateZoneGeometry(zone, width, height, geometryElapsedMs, safeRadius) {
    return {
        ...zone,
        ...zoneShape(width, height, safeRadius ?? 0, geometryElapsedMs, geometryElapsedMs),
        activeElapsedMs: geometryElapsedMs,
    };
}

function zoneShape(width, height, safeRadius, activeElapsedMs, geometryElapsedMs) {
    return {
        id: CLOSING_ZONE_ID,
        type: CLOSING_ZONE_TYPE,
        x: Number(width) / 2,
        y: Number(height) / 2,
        size: Math.max(0, Math.round(Number(safeRadius) * 2)),
        safeRadius: Math.max(0, Number(safeRadius) || 0),
        activeElapsedMs,
        geometryElapsedMs,
        visibility: "renderer-only",
        locked: true,
    };
}

function completedDamageTicks(previousElapsedMs, currentElapsedMs) {
    const start = Number(CLOSING_ZONE_CONFIG.startDelayMs);
    const interval = Math.max(1, Number(CLOSING_ZONE_CONFIG.damageIntervalMs) || 1);
    const previousActive = Math.max(0, previousElapsedMs - start);
    const currentActive = Math.max(0, currentElapsedMs - start);
    return Math.max(0, Math.floor(currentActive / interval) - Math.floor(previousActive / interval));
}

function fullArenaCircleRadius(width, height, config = CLOSING_ZONE_CONFIG) {
    return Math.hypot(Number(width) || 0, Number(height) || 0) / 2
        + Math.max(0, Number(config.initialSafeRadiusPaddingUnits) || 0);
}

function lerp(from, to, progress) {
    return from + (to - from) * Math.max(0, Math.min(1, progress));
}

File: server/src/main/java/com/example/botfight/simulation/ecs/entities/ClosingZoneSystem.java
package com.example.botfight.simulation.ecs.entities;

import com.example.botfight.simulation.gameconfig.ClosingZoneConfig;
import java.util.List;
import java.util.Set;

/**
 * Advances the authoritative arena-wide closing zone. It deliberately does
 * not implement an ability contract: the zone has no owner, target, or brain
 * payload and is only emitted as renderer/replay state.
 */
public final class ClosingZoneSystem {
    public static final String ID = "closing-zone";
    public static final String TYPE = "closingZone";

    private ClosingZoneSystem() {}

    public record State(double safeRadius, int activeElapsedMs, int geometryElapsedMs) {}

    public record TickResult<F extends AbilityEntityBot>(State state, ArenaEntity entity) {}

    public interface Damage<F extends AbilityEntityBot> {
        double maxHp(F bot);
        void apply(F bot, double amount);
    }

    public static <F extends AbilityEntityBot> TickResult<F> tick(
            State current,
            int elapsedMs,
            int stepMs,
            int width,
            int height,
            List<F> bots,
            ClosingZoneConfig config,
            Damage<F> damage) {
        if (elapsedMs < config.startDelayMs()) return new TickResult<>(null, null);

        int activeElapsedMs = Math.max(0, elapsedMs - config.startDelayMs());
        State initial = current == null
                ? new State(fullArenaCircleRadius(width, height, config), 0, 0)
                : current;
        int updateIntervalMs = Math.max(1, config.geometryUpdateMs());
        int geometryElapsedMs = activeElapsedMs / updateIntervalMs * updateIntervalMs;
        double candidateSafeRadius = safeRadiusAt(config.startDelayMs() + geometryElapsedMs, width, height, config);
        State next = Math.abs(candidateSafeRadius - initial.safeRadius()) > 1e-9
                ? new State(candidateSafeRadius,
                        activeElapsedMs, geometryElapsedMs)
                : initial;

        int previousElapsedMs = Math.max(0, elapsedMs - Math.max(1, stepMs));
        int damageTicks = completedDamageTicks(previousElapsedMs, elapsedMs, config);
        if (damage != null && damageTicks > 0) {
            for (int tick = 0; tick < damageTicks; tick += 1) {
                for (F bot : bots) {
                    if (!touchesDamageZone(next, bot, width, height)) continue;
                    double amount = Math.max(0, damage.maxHp(bot) * config.damagePercentMaxHp());
                    damage.apply(bot, amount);
                }
            }
        }
        return new TickResult<>(next, toEntity(next, width, height));
    }

    public static double safeRadiusAt(int elapsedMs, int width, int height, ClosingZoneConfig config) {
        if (elapsedMs < config.startDelayMs()) return Double.NaN;
        int activeElapsedMs = Math.max(0, elapsedMs - config.startDelayMs());
        double fullRadius = fullArenaCircleRadius(width, height, config);
        int phaseCount = Math.max(1, config.phaseCount());
        int approachMs = Math.max(0, config.approachDurationMs());
        int phaseMs = Math.max(1, config.phaseDurationMs());
        int completionElapsedMs = approachMs + Math.max(0, phaseCount - 1) * phaseMs;
        if (activeElapsedMs >= completionElapsedMs) return 0;
        int phaseIndex = activeElapsedMs / phaseMs;
        int phaseElapsedMs = activeElapsedMs % phaseMs;
        double startFraction = Math.max(0, 1.0 - (double) phaseIndex / phaseCount);
        if (phaseElapsedMs <= approachMs) {
            double progress = approachMs > 0 ? (double) phaseElapsedMs / approachMs : 1;
            double endFraction = Math.max(0, startFraction - 1.0 / phaseCount);
            return fullRadius * lerp(startFraction, endFraction, progress);
        }
        return fullRadius * Math.max(0, startFraction - 1.0 / phaseCount);
    }

    public static boolean touchesDamageZone(State zone, AbilityEntityBot bot, int width, int height) {
        if (zone == null || bot == null || bot.entityHp() <= 0) return false;
        double centerX = bot.entityX();
        double centerY = bot.entityY();
        double arenaCenterX = width / 2.0;
        double arenaCenterY = height / 2.0;
        return Math.hypot(centerX - arenaCenterX, centerY - arenaCenterY) + bot.entitySize() / 2.0
                >= zone.safeRadius();
    }

    private static ArenaEntity toEntity(State state, int width, int height) {
        return new ArenaEntity(
                ID, TYPE, 0, width / 2.0, height / 2.0,
                Math.max(0, (int) Math.round(state.safeRadius() * 2)),
                0, 0, 0, 0, true, 0, 0, 1.0, null, Set.of());
    }

    private static int completedDamageTicks(int previousElapsedMs, int elapsedMs, ClosingZoneConfig config) {
        int intervalMs = Math.max(1, config.damageIntervalMs());
        int previousActive = Math.max(0, previousElapsedMs - config.startDelayMs());
        int currentActive = Math.max(0, elapsedMs - config.startDelayMs());
        return Math.max(0, currentActive / intervalMs - previousActive / intervalMs);
    }

    private static double fullArenaCircleRadius(int width, int height, ClosingZoneConfig config) {
        return Math.hypot(width, height) / 2.0 + config.initialSafeRadiusPaddingUnits();
    }

    private static double lerp(double from, double to, double progress) {
        return from + (to - from) * Math.max(0, Math.min(1, progress));
    }
}
